core-reactive/src/main/java/io/micronaut/core/async/publisher/DelayedSubscriber.java

```diff
@@ -0,0 +1,149 @@
+    private Object completion;
+    private long demand;
+    private boolean cancel;
+
+    @Override
+    public void subscribe(Subscriber<? super T> s) {
+        s.onSubscribe(new Subscription() {
+            @Override
+            public void request(long n) {
+                synchronized (DelayedSubscriber.this) {
+                    demand = Math.max(demand + n, demand);
+                }
+                work();
+            }
+
+            @Override
+            public void cancel() {
+                synchronized (DelayedSubscriber.this) {
+                    cancel = true;
+                }
+                work();
+            }
+        });
+        synchronized (this) {
+            this.downstream = s;
+        }
+        work();
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        synchronized (this) {
+            this.upstream = s;
+        }
+        work();
+    }
+
+    @Override
+    public void onNext(T t) {
+        Subscriber<? super T> downstream = this.downstream;
+        if (downstream == null) {
+            throw new IllegalStateException("onNext before legitimate request");
+        }
+        downstream.onNext(t);
+    }
+
+    @Override
+    public void onError(Throwable t) {
+        synchronized (this) {
+            completion = t;
+        }
+        work();
+    }
+
+    @Override
+    public void onComplete() {
+        synchronized (this) {
+            completion = COMPLETE;
+        }
+        work();
+    }
+
+    private void work() {
+        boolean holdingWip = false;
+        while (true) {
+            Object completion = null;
+            boolean cancel = false;
+            long demand = 0;
+            synchronized (this) {
+                if (!holdingWip) {
+                    if (wip) {
+                        return;
+                    }
+                    wip = true;
+                    holdingWip = true;
+                }
+
+                if (this.completion != null && downstream != null) {
+                    completion = this.completion;
+                    this.completion = null;
+                } else if (this.demand != 0 && upstream != null && downstream != null) {
+                    demand = this.demand;
+                    this.demand = 0;
+                } else if (this.cancel && upstream != null) {
+                    cancel = true;
+                    this.cancel = false;
+                } else {
+                    // nothing to do
+                    wip = false;
+                    return;
+                }
+            }
+
+            if (completion != null) {
+                if (completion == COMPLETE) {
+                    downstream.onComplete();
+                } else {
+                    downstream.onError((Throwable) completion);
+                }
+            } else if (demand != 0) {
+                upstream.request(demand);
+            } else {
+                assert cancel;
```

Comment:
cancel is always true here because if none of the branches are hit, we would have returned above. intellij complained when i had `if (cancel)` here because it was able to figure out that cancel was always true, so i changed it to an assert.